### ISSUE TITLE
Built-in installer placeholders are displayed as format strings with [N] brackets

### ISSUE THREAD

**Issue by Funbit** [oleg-shilo/wixsharp#282]
Hello,

when I install/repair/uninstall an MSI file generated by the WixSharp, the UI displays some built-in placeholders as is, like when files are copying, for example:
![capture2](https://user-images.githubusercontent.com/5512433/35605107-d0aa4642-0689-11e8-9df8-470e84404722.PNG)

Is it a known behavior? Should I set some property to make the format strings to be evaluated?

Here is my environment:
- Windows 10 (Windows 8 works same way)
- WixSharp/WixSharp.bin: 1.6.1.0
- WiX Toolset: 3.11.0.1701
- Visual Studio 2017

and some non-default WiX project settings we use:
- new Property("MSIRESTARTMANAGERCONTROL", "Disable")
- new Property("MsiLogging", "wearclumpov!")
- project.InstallScope = InstallScope.perMachine
- project.UI = WUI.WixUI_Minimal
- project.ControlPanelInfo.NoModify = true

the rest settings are pretty basic...

Thanks in advance.

**Comment by JeremyAnsel** [comment]
Hello,
You can try that:
```csharp
            project.WixSourceGenerated += Project_WixSourceGenerated;

        private static void Project_WixSourceGenerated(System.Xml.Linq.XDocument document)
        {
            document.Root.Select("Product")
                .AddElement("UIRef", "Id=WixUI_ErrorProgressText");
        }

```

**Comment by Funbit** [comment]
Thank you, this reference solved the problem.
I also found this article:
https://www.firegiant.com/wix/tutorial/user-interface/ui-wizardry/
that states that the above reference is required to include WiX UI library.
Kind of a mess, but works.

**Comment by oleg-shilo** [comment]
In earlier releases WixSharp contained a proper formatting solution. Though after some reports of it exhibiting certain side efects the formatting was disabled uncionditionally.
I don't remember exactly the  actual nature of the side effects but it's possible to trace it with Git. Though it's not that important for your problem.
    
I suggest you create a WixSharp project from "Custom UI" template:
![image](https://user-images.githubusercontent.com/16729806/35608761-a137e3f4-06ae-11e8-971c-4a3ba242aa9e.png)

This will also add to the project a `ProgressDialog.cs` file.  This file contains `public override MessageResult ProcessMessage(...` routine where you will need to chage `simple` to false.

```C#
bool simple = true;
if (simple)
{
    for (int i = messageRecord.FieldCount - 1; i > 0; i--)
    {
        message = messageRecord[i].ToString();
    }
}
else
{
    message = messageRecord.FormatString;
    if (message.IsNotEmpty())
    {
        for (int i = 1; i < messageRecord.FieldCount; i++)
        {
            message = message.Replace("[" + i + "]", messageRecord[i].ToString());
        }
    }
    else
    {
        message = messageRecord[messageRecord.FieldCount - 1].ToString();
    }
}
```
You can also debug and see what exactly happening at runtime. Let me know if it fixed the problem and I may consider profigin a config value that user can use to control the formatting.

**Comment by Funbit** [comment]
Thank you for the explanation, Oleg. 
The Wix project file I'm using was created more than 2 years ago, I'm not sure what template was used back then. Anyway, I will try to recreate the project using the "Custom UI" template and see how it works.
Will get back here within a week or so.

**Comment by oleg-shilo** [comment]
Wait. I think there is some confusion here. 

I assumed you reported ManagedUI (WixSharp) problem. But it looks like you are dealing with native UI (WiX/MSI). 

If it is the case (and I think it is) then it's the same issue as #87 and #114.

**Comment by Funbit** [comment]
@oleg-shilo Yes, you are right :) It's a misunderstanding, my issue is exactly as described in #87... Missed it in my search results... So the issue is closed then :)

**Comment by oleg-shilo** [comment]
I re-categorized the issue as an enhancement. I think it is a good opportunity to improve Dev-experience and set `WixUI_ErrorProgressText` automatically.

**Comment by Funbit** [comment]
That would be a great enhancement indeed! Now it's impossible to figure this out without googling.

**Comment by oleg-shilo** [comment]
It's done now. 
Will be available in the very next release: https://github.com/oleg-shilo/wixsharp/blob/4f518908b6b29f070855c3ed11e66b0f821c618d/Source/src/WixSharp/Compiler.cs#L1354-L1355

**Comment by Funbit** [comment]
Oh, that was fast! Thank you again!
Looking forward to the release.

**Comment by Trigve** [comment]
I've been hit by this bug (using custom managed UI, but the progress dialog is the orignal wix#). I'm using 2.5.1. I've tried the suggestions above but nothing works. I want to add that I can reproduce it on Windows 11 only. On Windows 10 everything is working all right.

**Comment by oleg-shilo** [comment]
Hi @Trigve, while it's hard to give a comprehensive comment on the matter as progress message formatting model is one of the most convoluted is the most convoluted areas of WiX/MSI mix that is already convoluted enough. 
The most comprehensive discussion on the subject has happened here: https://github.com/oleg-shilo/wixsharp/discussions/1504

The essence of the limitation is directly captured in the code:

https://github.com/oleg-shilo/wixsharp/blob/88ce6d33b58064153a73875094ce052dae250bb4/Source/src/WixSharp.UI/ManagedUI/Forms/ProgressDialog.cs#L130 

To put it simply, the implementation of MSI progress message formatting is fragile and often does not work. At run time MSI may not not supply the values for the substitution in the formatting template string.

Windows 11 vs 10 is interesting though. I have never heard about the difference. 
Of course, since it works on 10 it means that your msi is valid and the implementation of the MSI Windows service causes the difference in behaviour.

However this seems like a thing that could be noticed by many so quite possibly there is some WiX work around that simply avoids the problem instead of fixing it. I suggest you check some WiX forums. If you can find the solution there bringing it wo WixSharop is easy and I am happy to help you with that.

**Comment by oleg-shilo** [comment]
Hm, I just tested it and I see that it does inject this element. 
It's all controlled by `project.LocalizeErrorAndProgressText` as in some edge cases you may want to avoid this element. But by default, it is set to true and inserted.

I tested InstallDir sample by commenting out `project.LocalizeErrorAndProgressText = false;` and the required injection does appear in the wxs file:

![Image](https://github.com/user-attachments/assets/08aab297-ed16-43f5-8b45-2a98bd28e45d)

Do you see it in your project?

**Comment by oleg-shilo** [comment]
Fantastic, thank you very much for going the extra mile. 
Indeed the very second XML generation step removes all `UIRef` elements in case the setup is using ManagedUI (your case). This was done because this element was conflicting with EmbeddedUI in WiX3 and this constraint, apparently, is no longer present in WiX4 schema. So I corrected it and the next release will have this problem fixed.

The problem with my test that showed the generation of the `UIRef` element was caused by the use of the native MSI UI. Effectively my and your tests were not using the same UI model. That was the problem.

Thank you for your help with fixing this.
